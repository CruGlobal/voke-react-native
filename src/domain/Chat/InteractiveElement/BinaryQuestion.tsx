import React from 'react';
import { TMessage } from 'utils/types';
import { StyleSheet, View } from 'react-native';
import Image from 'components/Image';
import Text from 'components/Text';
import LinearGradient from 'react-native-linear-gradient';
import Button from 'components/Button';

import styles from './styles';

interface Props {
  metadata: TMessage['metadata'];
  onValueSelected: (value: string) => void;
}

const BinaryQuestion = ({
  metadata,
  onValueSelected,
}: Props): React.ReactElement => {
  const image = metadata?.image?.small || '';
  const name = metadata?.name || '';
  const comment = metadata?.comment || '';
  const question = metadata?.question || '';
  const answers = metadata?.answers || [];
  const hasSelected = answers.find(a => a.selected) ? true : false;
  return (
    <View style={styles.binaryBubble}>
      {image ? (
        <Image source={{ uri: image }} style={StyleSheet.absoluteFill} />
      ) : null}
      <Text style={styles.binaryName}>{name}</Text>
      <Text style={styles.binaryComment}>{comment}</Text>
      <LinearGradient
        colors={['rgba(0,0,0,0)', 'rgba(0,0,0,5)']}
        style={styles.mainQuestion}
      >
        <Text style={styles.binaryQuestion}>{question}</Text>
        <View style={styles.binaryOptionsWrap}>
          {answers.map((a, index) => (
            <Button
              key={index}
              disabled={hasSelected}
              onPress={(): void => {
                onValueSelected(a.value);
              }}
              color={a.selected ? 'blank' : 'accent'}
              style={[
                a.selected || !hasSelected ? { opacity: 1 } : { opacity: 0.4 },
              ]}
              testID={'binaryButton-' + index}
            >
              {a.key}
            </Button>
          ))}
        </View>
      </LinearGradient>
    </View>
  );
};

export default BinaryQuestion;
