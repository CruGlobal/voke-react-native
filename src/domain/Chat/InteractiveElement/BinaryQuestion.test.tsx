import React from 'react';
import { fireEvent } from '@testing-library/react-native';
import { Host } from 'react-native-portalize';
// import { toHaveStyle } from '@testing-library/jest-native'; - causes TS error.
// https://github.com/testing-library/jest-dom/issues/123
import '@testing-library/jest-native';
import {
  groupAdventureMock,
  stepMock,
  mockMyMessage,
  soloAdventureMock,
  botMessageMock,
  mockUser,
  mockMessage,
  mockStepDecisionInactive,
  mockMessageBinaryDefault,
  mockMessageBinarySelected,
} from 'mocks/vokeDataMocks';
import renderWithContext from 'utils/testUtils';

import BinaryQuestion from './BinaryQuestion';

const initialState = {
  auth: { user: mockUser },
  data: {
    adventureStepMessages: {
      [mockStepDecisionInactive.id]: {},
    },
  },
};

beforeEach(() => {
  // https://github.com/facebook/jest/issues/6434#issuecomment-525576660
  jest.useFakeTimers();
});

it('renders correctly binary question - default', () => {
  const { toJSON } = renderWithContext(
    <BinaryQuestion
      metadata={mockMessageBinaryDefault.metadata}
      onValueSelected={(): void => {
        //void
      }}
    />,
    { initialState },
  );

  expect(toJSON()).toMatchSnapshot();
});

it('renders correctly binary question - selected', () => {
  const { toJSON } = renderWithContext(
    <BinaryQuestion
      metadata={mockMessageBinarySelected.metadata}
      onValueSelected={(): void => {
        //void
      }}
    />,
    { initialState },
  );

  expect(toJSON()).toMatchSnapshot();
});

it('renders correctly binary question - partial data', () => {
  const { toJSON } = renderWithContext(
    <BinaryQuestion
      metadata={{
        messenger_journey_step_id: '1693da4d-a8bd-40e3-99fd-913a1c8df372',
        step_kind: 'binary',
        vokebot_action: 'journey_step',
      }}
      onValueSelected={(): void => {
        //void
      }}
    />,
    { initialState },
  );

  expect(toJSON()).toMatchSnapshot();
});

it('emit events correctly for binary question', () => {
  const onValueSelected = jest.fn();
  const { getByTestId } = renderWithContext(
    <BinaryQuestion
      metadata={mockMessageBinaryDefault.metadata}
      onValueSelected={(val): void => {
        onValueSelected(val);
      }}
    />,
    { initialState },
  );

  fireEvent.press(getByTestId('binaryButton-0'));
  expect(onValueSelected).toHaveBeenCalledWith(
    '26cbf173-bb2f-4ccf-b3c9-11ed98733790',
  );
  onValueSelected.mockClear();

  fireEvent.press(getByTestId('binaryButton-1'));
  expect(onValueSelected).toHaveBeenCalledWith(
    'fa9bd7f8-4016-4c10-b7ec-70871f74bcc6',
  );
});

it('not emiting events in binary question when already selected', () => {
  const onValueSelected = jest.fn();
  const { getByTestId } = renderWithContext(
    <BinaryQuestion
      metadata={mockMessageBinarySelected.metadata}
      onValueSelected={(val): void => {
        onValueSelected(val);
      }}
    />,
    { initialState },
  );

  fireEvent.press(getByTestId('binaryButton-0'));
  expect(onValueSelected).not.toHaveBeenCalled();
});
