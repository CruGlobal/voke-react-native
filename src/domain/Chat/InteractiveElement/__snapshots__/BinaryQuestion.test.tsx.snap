// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly binary question - default 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#1d1d26",
      "borderRadius": 10,
      "overflow": "hidden",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
      ]
    }
  >
    <FastImageView
      resizeMode="cover"
      source={
        Object {
          "uri": "https://cru-vokeapi-prod.s3.amazonaws.com/platform/organization/journey/steps/images/811/1ed/6d-/small/data.jpg?1581968878",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#1d1d26",
        },
        Object {
          "fontFamily": "TitilliumWeb-Regular",
        },
        Object {
          "color": "#fff",
          "fontSize": 21,
          "paddingBottom": 8,
          "paddingHorizontal": 22,
          "paddingTop": 22,
        },
      ]
    }
  >
    PRAYER
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#1d1d26",
        },
        Object {
          "fontFamily": "TitilliumWeb-Regular",
        },
        Object {
          "color": "#fff",
          "fontSize": 17,
          "paddingHorizontal": 22,
        },
      ]
    }
  >
    Jesus, thank you for loving me, thank you for the worth you speak into
  </Text>
  <BVLinearGradient
    colors={
      Array [
        0,
        4278190080,
      ]
    }
    endPoint={
      Object {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={null}
    startPoint={
      Object {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      Object {
        "paddingBottom": 44,
        "paddingHorizontal": 16,
        "paddingTop": 22,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1d1d26",
          },
          Object {
            "fontFamily": "TitilliumWeb-Regular",
          },
          Object {
            "color": "#fff",
            "flex": 1,
            "fontSize": 21,
            "lineHeight": 27.3,
            "textAlign": "center",
          },
        ]
      }
    >
      Have you made this commitment today?
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-around",
          "paddingHorizontal": 22,
          "paddingTop": 16,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#FF9900",
            "borderRadius": 88,
            "opacity": 1,
            "paddingHorizontal": 16,
            "paddingVertical": 8,
          }
        }
        testID="binaryButton-0"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "TitilliumWeb-Regular",
                  "textAlign": "center",
                },
                Object {
                  "fontSize": 21,
                  "lineHeight": 31.5,
                },
                Object {
                  "color": "#fff",
                },
                undefined,
                undefined,
              ]
            }
          >
            Not yet
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#FF9900",
            "borderRadius": 88,
            "opacity": 1,
            "paddingHorizontal": 16,
            "paddingVertical": 8,
          }
        }
        testID="binaryButton-1"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "TitilliumWeb-Regular",
                  "textAlign": "center",
                },
                Object {
                  "fontSize": 21,
                  "lineHeight": 31.5,
                },
                Object {
                  "color": "#fff",
                },
                undefined,
                undefined,
              ]
            }
          >
            I just did
          </Text>
        </View>
      </View>
    </View>
  </BVLinearGradient>
</View>
`;

exports[`renders correctly binary question - partial data 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#1d1d26",
      "borderRadius": 10,
      "overflow": "hidden",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#1d1d26",
        },
        Object {
          "fontFamily": "TitilliumWeb-Regular",
        },
        Object {
          "color": "#fff",
          "fontSize": 21,
          "paddingBottom": 8,
          "paddingHorizontal": 22,
          "paddingTop": 22,
        },
      ]
    }
  >
    
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#1d1d26",
        },
        Object {
          "fontFamily": "TitilliumWeb-Regular",
        },
        Object {
          "color": "#fff",
          "fontSize": 17,
          "paddingHorizontal": 22,
        },
      ]
    }
  >
    
  </Text>
  <BVLinearGradient
    colors={
      Array [
        0,
        4278190080,
      ]
    }
    endPoint={
      Object {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={null}
    startPoint={
      Object {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      Object {
        "paddingBottom": 44,
        "paddingHorizontal": 16,
        "paddingTop": 22,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1d1d26",
          },
          Object {
            "fontFamily": "TitilliumWeb-Regular",
          },
          Object {
            "color": "#fff",
            "flex": 1,
            "fontSize": 21,
            "lineHeight": 27.3,
            "textAlign": "center",
          },
        ]
      }
    >
      
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-around",
          "paddingHorizontal": 22,
          "paddingTop": 16,
        }
      }
    />
  </BVLinearGradient>
</View>
`;

exports[`renders correctly binary question - selected 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#1d1d26",
      "borderRadius": 10,
      "overflow": "hidden",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
      ]
    }
  >
    <FastImageView
      resizeMode="cover"
      source={
        Object {
          "uri": "https://cru-vokeapi-prod.s3.amazonaws.com/platform/organization/journey/steps/images/811/1ed/6d-/small/data.jpg?1581968878",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#1d1d26",
        },
        Object {
          "fontFamily": "TitilliumWeb-Regular",
        },
        Object {
          "color": "#fff",
          "fontSize": 21,
          "paddingBottom": 8,
          "paddingHorizontal": 22,
          "paddingTop": 22,
        },
      ]
    }
  >
    PRAYER
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#1d1d26",
        },
        Object {
          "fontFamily": "TitilliumWeb-Regular",
        },
        Object {
          "color": "#fff",
          "fontSize": 17,
          "paddingHorizontal": 22,
        },
      ]
    }
  >
    Jesus, thank you for loving me, thank you for the worth you speak into
  </Text>
  <BVLinearGradient
    colors={
      Array [
        0,
        4278190080,
      ]
    }
    endPoint={
      Object {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={null}
    startPoint={
      Object {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      Object {
        "paddingBottom": 44,
        "paddingHorizontal": 16,
        "paddingTop": 22,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1d1d26",
          },
          Object {
            "fontFamily": "TitilliumWeb-Regular",
          },
          Object {
            "color": "#fff",
            "flex": 1,
            "fontSize": 21,
            "lineHeight": 27.3,
            "textAlign": "center",
          },
        ]
      }
    >
      Have you made this commitment today?
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-around",
          "paddingHorizontal": 22,
          "paddingTop": 16,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#fff",
            "borderRadius": 88,
            "opacity": 1,
            "paddingHorizontal": 16,
            "paddingVertical": 8,
          }
        }
        testID="binaryButton-0"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "TitilliumWeb-Regular",
                  "textAlign": "center",
                },
                Object {
                  "fontSize": 21,
                  "lineHeight": 31.5,
                },
                Object {
                  "color": "#186078",
                },
                undefined,
                undefined,
              ]
            }
          >
            Not yet
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#FF9900",
            "borderRadius": 88,
            "opacity": 1,
            "paddingHorizontal": 16,
            "paddingVertical": 8,
          }
        }
        testID="binaryButton-1"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "TitilliumWeb-Regular",
                  "textAlign": "center",
                },
                Object {
                  "fontSize": 21,
                  "lineHeight": 31.5,
                },
                Object {
                  "color": "#fff",
                },
                undefined,
                undefined,
              ]
            }
          >
            I just did
          </Text>
        </View>
      </View>
    </View>
  </BVLinearGradient>
</View>
`;
